solidity/contracts/ECDSARewardsDistributor.sol

```diff
@@ -0,0 +1,109 @@
+/// - added receiveApproval() function that will be called each time to allocate
+///   new KEEP rewards for a given merkle root. Merkle root is going to be generated
+///   regulary (ex. every week) and it is also means that an interval for that
+///   merkle root has passed
+/// - changed code accordingly to process claimed rewards using a map of merkle
+///   roots
+contract ECDSARewardsDistributor is Ownable {
+    using SafeERC20 for KeepToken;
+    using BytesLib for bytes;
+
+    KeepToken public token;
+
+    // This event is triggered whenever a call to #claim succeeds.
+    event RewardsClaimed(uint256 index, address account, uint256 amount);
+    // This event is triggered whenever rewards are allocated.
+    event RewardsAllocated(bytes32 merkleRoot, uint256 amount);
+
+    // Merkle root -> total amount for distribution for a given interval.
+    mapping(bytes32 => uint256) private merkleRoots;
+    // Bytes32 key is a merkle root and the value is a packed array of booleans.
+    mapping(bytes32 => mapping(uint256 => uint256)) private claimedBitMap;
+
+    constructor(address _token) public {
+        token = KeepToken(_token);
+    }
+
+    /// Call receiveApproval to allocate amount of KEEP for a given merkle root.
+    /// @param _from The original sender of the tokens.
+    /// @param _amount The amount of KEEP tokens to fund.
+    /// @param _token The KEEP token to fund the rewards in.
+    /// @param _extraData Merkle root (32 bytes).
+    function receiveApproval(
+        address _from,
+        uint256 _amount,
+        address _token,
+        bytes memory _extraData
+    ) public {
+        require(IERC20(_token) == token, "Unsupported token");
+        require(_amount <= token.balanceOf(_from), "Sender must have enough amount");
+        require(_extraData.length == 32, "Wrong lenght of merkle root");
+
+        token.safeTransferFrom(_from, address(this), _amount);
+
+        bytes32 merkleRoot = _extraData.toBytes32();
+
+        merkleRoots[merkleRoot] = _amount;
+
+        emit RewardsAllocated(merkleRoot, _amount);
+    }
+
+    function isClaimed(uint256 index, bytes32 merkleRoot) public view returns (bool) {
+        uint256 claimedWordIndex = index / 256;
+        uint256 claimedBitIndex = index % 256;
+        uint256 claimedWord = claimedBitMap[merkleRoot][claimedWordIndex];
+        uint256 mask = (1 << claimedBitIndex);
+        return claimedWord & mask == mask;
+    }
+
+    function _setClaimed(uint256 index, bytes32 merkleRoot) private {
+        uint256 claimedWordIndex = index / 256;
+        uint256 claimedBitIndex = index % 256;
+        claimedBitMap[merkleRoot][claimedWordIndex] = claimedBitMap[merkleRoot][claimedWordIndex] | (1 << claimedBitIndex);
+    }
+
+    function claim(uint256 index, address account, uint256 amount, bytes32 merkleRoot, bytes32[] calldata merkleProof) external {
+        require(!isClaimed(index, merkleRoot), 'Rewards already claimed');
```

Comment:
`Rewards` or `Reward`?

Reply:
I think both work actually https://www.usingenglish.com/forum/threads/27257-reward-vs-rewards depending on how you look at it.

Anyway, I can change it to `Reward`.